File: DSA/polynomial.c
#include <stdio.h>
#include<math.h>
// addition function
void add(int a[], int b[]){
     int i;
     int c[10];
     for(i = 0; i < 4; i++){
        c[i] = a[i] + b[i];
     }
    for (i = 0; i < 4; i++) {
        printf("%dx^%d", c[i], i);
        if (i < 3) {
            printf(" + ");
        }
    }
    printf("\n");    
}
// difference function
void sub(int a[], int b[]){
     int i;
     int c[10];
     for(i = 0; i < 4; i++){
        c[i] = a[i] - b[i];
     }
    for (i = 0; i < 4; i++) {
        printf("%dx^%d", c[i], i);
        if (i < 3) {
            printf(" + ");
        }
    }
    printf("\n");  
}
// multiplication function
void multiply(int a[], int b[]){
     int i,j;
     int c[7] = {0};
     for(i = 0; i < 4; i++){
        for(j =0; j < 4; j++){
        c[i + j] += b[j] * a[i];
        }
     }
    for (i = 0; i < 7; i++) {
        printf("%dx^%d", c[i], i);
        if (i < 6) {
            printf(" + ");
        }
    }
    printf("\n"); 
}
//evaluate function
void evaluate(int a[],int val){
    int res = 0,i;
    for(i = 0; i < 4; i++){
        res += pow(2,i) * a[i];
    }
    printf("f(%d): %d",val,res);
}
void derivative(int a[]){
    int c[4] = {0},i;
    for(i = 0; i < 4; i++){
        c[i] = a[i + 1] * (i + 1);
    }
    printf("\n");
    for (i = 0; i < 3; i++) {
        printf("%dx^%d", c[i], i);
        if (i < 2) {
            printf(" + ");
        }
    }
}
int main() {
    //polynomial creation
    int a[10] = {1, 2, 3, 1}, b[10] = {2, 3, 4, 2};
    int i;
    for (i = 0; i < 4; i++) {
        printf("%dx^%d", a[i], i);
        if (i < 3) {
            printf(" + ");
        }
    }
    printf("\n");
    for (i = 0; i < 4; i++) {
        printf("%dx^%d", b[i], i);
        if (i < 3) {
            printf(" + ");
        }
    }
    //function calls
    printf("\n");
    add(a,b);
    sub(a,b);
    multiply(a,b);
    evaluate(a,2);
    derivative(a);
    return 0;
}

File: MSC_JAVA/polynomial/src/Polynomial.java
/**
 * This class will create any arbitrary polynomial to work with. It uses linked link to 
 * store the terms of the polynomial.
 *  
 * @author Sunirmal Khatua
 *
 */
public class Polynomial {
    // ********** attributes ********************
	private Node head;
	private int degree;
	
    /*
     * Default constructor
     */
    public Polynomial() {
    	//TO BE IMPLEMENTED
        this.head = null;
        this.degree = 0;
    }
    
    /*
     * Parameterized constructor
     */
    public Polynomial(double[] coeff) {           
    	//TO BE IMPLEMENTED
        for(int i = 0; i < coeff.length; i++){
            insertSorted(coeff[i], i);
        }
        this.degree = coeff.length - 1;
    }
    
    /*
     * This is a utility method that can be used to insert a term in sorted order
     * in the current polynomial 
     */
    private void insertSorted(double coeff, int power) {   	
    	if(head==null) {
    		head = new Node(new Terms(coeff, power), null);
    	}else {
	    	Node curr = head.next;
	    	Node prev = head;
	    	while(curr != null && curr.term.power<power) {
	    		prev = curr;
	    		curr = curr.next;
	    	}
	    	if(curr!=null && curr.term.power == power) 
	    		curr.term.coeff += coeff;
	    	else
	    		prev.next = new Node(new Terms(coeff, power), curr);
    	}
    }
    
    /*
     *  subtract a polynomial q from the given polynomial p i.e.  p = p-q
     */
    public void subtract(Polynomial p) {
    	//TO BE IMPLEMENTED
    }
    
    /*
     * returns value of p(x)
     */
    public double evaluate(double x) {
        double res = 0.0;
        Node temp = this.head;
        while (temp != null) {
            res += temp.term.coeff * (Math.pow(x,temp.term.power));
            temp = temp.next;
        }
    	return res;
    }
    
    /*
     *  add a polynomial q to the given polynomial p i.e. p = p+q
     */
    public void add(Polynomial p) {
    	degree = Math.max(degree, p.degree); 
    	head = sum(this, p).head;
    }
    
     /*
      *  multiply the given polynomial by a polynomial q i.e.  p = p*q
      */
    public void multiply(Polynomial q) {
    	     degree = this.degree+q.degree;
    	     head = product(this, q).head;
    	     
    }
    
     /*
      *  multiply the given polynomial by a constant a  i.e   p = a.p
      */
    public void scale(double a) {
    	//TO BE IMPLEMENTED
        Node temp = this.head;
        while (temp != null) {
            temp.term.coeff = temp.term.coeff * a;
            temp = temp.next;
        }
    }
    
    /*
     * prints a polynomial object
     */
    public String toString() {
            Node curr = head;
            String q = "";
            if(curr!=null) {
            	if(curr.term.power == 1)
            		q=curr.term.coeff + "x";
            	else if (curr.term.power == 0)
            		q = ""+curr.term.coeff;
            	else
            		q=curr.term.coeff + "x^" + curr.term.power;
            	curr = curr.next;
            }
            while(curr != null) {
            	if(curr.term.power == 1)
            		q=curr.term.coeff + "x"+"+"+q;
            	else
            		q = curr.term.coeff + "x^" + curr.term.power+"+"+q; 
            	curr = curr.next;
            } 
            return q;
    }
    
    /*
     * multiplies 2 polynomials  and creates a new polynomial 
     * without destroying the original i.e. r = p*q
     */
    public static Polynomial product(Polynomial p, Polynomial q) {
    	Polynomial temp = new Polynomial();
    	//TO BE IMPLEMENTED
        double coeff = 0.0;
        int power = p.head.term.power;
        Node a = p.head;
        Node b = q.head;
        Node c = temp.head;
        while(a != null){
            while(b != null){
            if(temp.head == null){
              c = new Node(new Terms(coeff, power), null);
              temp.head = c;
            }
            else{
                Node prev = temp.head;
                while(prev.next != null){
                    prev = prev.next;
                }
                c = new Node(new Terms(coeff, power), null);
                prev.next = c;
            }
            c.term.coeff = a.term.coeff * b.term.coeff;
            c.term.power = a.term.power + b.term.power;
            b = b.next;
        }
        b = q.head;
        a = a.next;
    }
        temp.degree = p.degree;
        return temp;
    }
    
    /*
     *  adds 2 polynomials  and creates a new polynomial without destroying 
     *  the original i.e.  r = p+q
     */
    public static Polynomial sum(Polynomial p, Polynomial q) {
    	Polynomial temp = new Polynomial();
    	//TO BE IMPLEMENTED
        double coeff = 0.0;
        int power = p.head.term.power;
        Node a = p.head;
        Node b = q.head;
        Node c = temp.head;
        while(a != null && b != null){
            if(temp.head == null){
              c = new Node(new Terms(coeff, power), null);
              temp.head = c;
            }
            else{
                Node prev = temp.head;
                while(prev.next != null){
                    prev = prev.next;
                }
                c = new Node(new Terms(coeff, power), null);
                prev.next = c;
            }
            c.term.coeff = a.term.coeff + b.term.coeff;
            c.term.power = a.term.power;
            a = a.next;
            b = b.next;
        }
        temp.degree = p.degree;
        return temp;
    }
    
    /*
     *  subtracts 2 polynomials  and creates a new polynomial without destroying 
     *  the originals  i.e. r = p-q
     */
    public static Polynomial diff(Polynomial p, Polynomial q) {
    	Polynomial temp = new Polynomial();
    	//TO BE IMPLEMENTED
        double coeff = 0.0;
        int power = p.head.term.power;
        Node a = p.head;
        Node b = q.head;
        Node c = temp.head;
        while(a != null && b != null){
            if(temp.head == null){
              c = new Node(new Terms(coeff, power), null);
              temp.head = c;
            }
            else{
                Node prev = temp.head;
                while(prev.next != null){
                    prev = prev.next;
                }
                c = new Node(new Terms(coeff, power), null);
                prev.next = c;
            }
            c.term.coeff = a.term.coeff - b.term.coeff;
            c.term.power = a.term.power;
            a = a.next;
            b = b.next;
        }
        temp.degree = p.degree;
        return temp;
    }
    

}
